File: app/src/main/overlays/issues-overlay.ts
import type { OverlayManager, OverlayItem } from '../overlay-manager';
import type { TrackerService, Issue } from '../services/tracker';
import type { ClaudeSession } from '../claude-session';
import type { StatusBarState } from '../status-bar-state';
import type { Config } from '../config';
import { showTrackerSetup } from './tracker-setup';

export function showIssuesOverlay(
  overlay: OverlayManager,
  tracker: TrackerService,
  session: ClaudeSession,
  statusBar: StatusBarState,
  config: Config,
  onTrackerRefresh: () => void,
  onUpdate: () => void,
): void {
  if (!tracker.enabled) {
    overlay.show({
      title: 'Issues',
      items: [
        { label: 'No tracker configured', dimmed: true },
        { label: '', dimmed: true },
        {
          label: '[C] Configure Tracker',
          shortcut: 'c',
          action: () => showTrackerSetup(overlay, config, () => {
            onTrackerRefresh();
            onUpdate();
          }, onUpdate),
        },
      ],
      onClose: onUpdate,
      footer: '[C] Configure',
    });
    return;
  }

  // Show loading state immediately
  overlay.show({
    title: `Issues (${tracker.type}: ${tracker.teamKey})`,
    items: [{ label: 'Loading issues...', dimmed: true }],
    onClose: onUpdate,
    width: 70,
    footer: '[Enter] Select  [A] Analyze  [P] Plan  [X] Fix  [N] New  [S] Status',
  });

  // Fetch issues
  loadIssues(overlay, tracker, session, statusBar, config, onTrackerRefresh, onUpdate);
}

async function loadIssues(
  overlay: OverlayManager,
  tracker: TrackerService,
  session: ClaudeSession,
  statusBar: StatusBarState,
  config: Config,
  onTrackerRefresh: () => void,
  onUpdate: () => void,
): Promise<void> {
  try {
    overlay.setLoading(true);
    const issues = await tracker.listIssues({ limit: 20 });
    overlay.setLoading(false);

    if (issues.length === 0) {
      overlay.updateItems([
        { label: 'No open issues found', dimmed: true },
        { label: '', dimmed: true },
        { label: '[N] Create New Issue', shortcut: 'n', action: () => promptNewIssue(overlay, tracker, onUpdate) },
        { label: '', dimmed: true },
        {
          label: '[C] Reconfigure Tracker',
          shortcut: 'c',
          action: () => showTrackerSetup(overlay, config, () => { onTrackerRefresh(); onUpdate(); }, onUpdate),
        },
      ]);
      return;
    }

    const items: OverlayItem[] = [];

    // Current issue indicator
    if (statusBar.currentIssue) {
      items.push({
        label: `Current: ${statusBar.currentIssue.key} \u2014 ${statusBar.currentIssue.title}`,
        value: statusBar.currentIssue.status,
        dimmed: true,
      });
      items.push({ label: '', dimmed: true });
    }

    // Issue list
    for (const issue of issues) {
      const isCurrent = statusBar.currentIssue?.key === issue.key;
      const priorityIcon = priorityToIcon(issue.priority);
      items.push({
        label: `${isCurrent ? '\u25c6' : ' '} ${issue.key}  ${priorityIcon} ${truncate(issue.title, 40)}`,
        value: issue.status,
        action: () => selectIssue(overlay, tracker, session, statusBar, issue, config, onTrackerRefresh, onUpdate),
      });
    }

    items.push({ label: '', dimmed: true });
    items.push({
      label: '[N] Create New Issue',
      shortcut: 'n',
      action: () => promptNewIssue(overlay, tracker, onUpdate),
    });
    items.push({
      label: '[C] Reconfigure Tracker',
      shortcut: 'c',
      action: () => showTrackerSetup(overlay, config, () => { onTrackerRefresh(); onUpdate(); }, onUpdate),
    });

    overlay.updateItems(items);
  } catch (err: any) {
    overlay.setLoading(false);
    overlay.updateItems([
      { label: `Error: ${err.message}`, dimmed: true },
      { label: '', dimmed: true },
      { label: '[R] Retry', shortcut: 'r', action: () => loadIssues(overlay, tracker, session, statusBar, config, onTrackerRefresh, onUpdate) },
      {
        label: '[C] Reconfigure Tracker',
        shortcut: 'c',
        action: () => showTrackerSetup(overlay, config, () => { onTrackerRefresh(); onUpdate(); }, onUpdate),
      },
    ]);
  }
}

function selectIssue(
  overlay: OverlayManager,
  tracker: TrackerService,
  session: ClaudeSession,
  statusBar: StatusBarState,
  issue: Issue,
  config: Config,
  onTrackerRefresh: () => void,
  onUpdate: () => void,
): void {
  // Show issue action submenu
  overlay.updateItems([
    { label: `${issue.key}: ${issue.title}`, dimmed: true },
    { label: `Status: ${issue.status}  Priority: ${issue.priority}`, dimmed: true },
    { label: '', dimmed: true },
    {
      label: '[C] Set as Current Issue',
      shortcut: 'c',
      action: () => {
        statusBar.currentIssue = issue;
        onUpdate();
        overlay.showMessage(`Set current: ${issue.key}`);
      },
    },
    {
      label: '[A] Analyze Issue',
      shortcut: 'a',
      action: () => {
        statusBar.currentIssue = issue;
        session.write(`Analyze this issue and tell me what you think the problem is and what files are involved:\n${issue.key}: ${issue.title}\n`);
        overlay.close();
      },
    },
    {
      label: '[P] Plan Implementation',
      shortcut: 'p',
      action: () => {
        statusBar.currentIssue = issue;
        session.write(`Create an implementation plan for this issue. List the files to change and the approach:\n${issue.key}: ${issue.title}\n`);
        overlay.close();
      },
    },
    {
      label: '[X] Fix Issue',
      shortcut: 'x',
      action: () => {
        statusBar.currentIssue = issue;
        session.write(`Fix this issue. Analyze it, plan the fix, implement it, and test it:\n${issue.key}: ${issue.title}\n`);
        overlay.close();
      },
    },
    {
      label: '[S] Update Status',
      shortcut: 's',
      action: () => showStatusPicker(overlay, tracker, issue, onUpdate),
    },
    {
      label: '[O] Open in Browser',
      shortcut: 'o',
      action: () => {
        import('child_process').then(({ exec }) => exec(`open "${issue.url}"`));
        overlay.showMessage('Opening in browser...');
      },
    },
    { label: '', dimmed: true },
    {
      label: '[B] \u2190 Back to List',
      shortcut: 'b',
      action: () => loadIssues(overlay, tracker, session, statusBar, config, onTrackerRefresh, onUpdate),
    },
  ]);
}

function showStatusPicker(
  overlay: OverlayManager,
  tracker: TrackerService,
  issue: Issue,
  onUpdate: () => void,
): void {
  const statuses = tracker.type === 'linear'
    ? ['Backlog', 'Todo', 'In Progress', 'In Review', 'Done', 'Canceled']
    : ['To Do', 'In Progress', 'In Review', 'Done'];

  overlay.updateItems(
    statuses.map((s) => ({
      label: `${s === issue.status ? '\u25c6 ' : '  '}${s}`,
      action: async () => {
        overlay.setLoading(true);
        try {
          await tracker.updateStatus(issue.key, s);
          issue.status = s;
          overlay.showMessage(`Updated to ${s}`);
          onUpdate();
        } catch (err: any) {
          overlay.showMessage(`Error: ${err.message}`);
        }
        overlay.setLoading(false);
      },
    })),
  );
}

function promptNewIssue(
  _overlay: OverlayManager,
  _tracker: TrackerService,
  _onUpdate: () => void,
): void {
  _overlay.showMessage('Use Claude to create issues (coming soon)');
}

function priorityToIcon(priority: string): string {
  const p = priority.toLowerCase();
  if (p.includes('urgent') || p.includes('critical')) return '\u25a0\u25a0\u25a0';
  if (p.includes('high')) return '\u25a0\u25a0\u25a1';
  if (p.includes('medium') || p.includes('med')) return '\u25a0\u25a1\u25a1';
  if (p.includes('low')) return '\u25a1\u25a1\u25a1';
  return '\u00b7\u00b7\u00b7';
}

function truncate(str: string, max: number): string {
  return str.length > max ? str.slice(0, max - 1) + '\u2026' : str;
}

File: app-native/Sources/Orch/Overlays/IssuesOverlay.swift
import Foundation

func showIssuesOverlay(
    overlay: OverlayManager,
    tracker: TrackerService,
    session: ClaudeSession,
    statusBar: StatusBarState,
    config: Config,
    onTrackerRefresh: @escaping () -> Void,
    onUpdate: @escaping () -> Void
) {
    guard tracker.enabled else {
        overlay.show(OverlayConfig(
            title: "Issues",
            items: [
                OverlayItem(label: "No tracker configured", dimmed: true),
                OverlayItem(label: "", dimmed: true),
                OverlayItem(label: "[C] Configure Tracker", shortcut: "c", action: {
                    showTrackerSetup(overlay: overlay, config: config, onConfigured: { onTrackerRefresh(); onUpdate() }, onUpdate: onUpdate)
                }),
            ],
            onClose: onUpdate,
            footer: "[C] Configure",
            anchor: "status-issues"
        ))
        return
    }

    overlay.show(OverlayConfig(
        title: "Issues (\(tracker.type): \(tracker.teamKey))",
        items: [OverlayItem(label: "Loading issues...", dimmed: true)],
        onClose: onUpdate,
        width: 70,
        footer: "[Enter] Select  [A] Analyze  [P] Plan  [X] Fix  [N] New  [S] Status",
        anchor: "status-issues"
    ))

    loadIssues(overlay: overlay, tracker: tracker, session: session, statusBar: statusBar, config: config, onTrackerRefresh: onTrackerRefresh, onUpdate: onUpdate)
}

private func loadIssues(
    overlay: OverlayManager,
    tracker: TrackerService,
    session: ClaudeSession,
    statusBar: StatusBarState,
    config: Config,
    onTrackerRefresh: @escaping () -> Void,
    onUpdate: @escaping () -> Void
) {
    overlay.setLoading(true)
    Task {
        do {
            let issues = try await tracker.listIssues(status: nil, assignedToMe: false, limit: 20)
            await MainActor.run {
                overlay.setLoading(false)

                if issues.isEmpty {
                    overlay.updateItems([
                        OverlayItem(label: "No open issues found", dimmed: true),
                        OverlayItem(label: "", dimmed: true),
                        OverlayItem(label: "[N] Create New Issue", shortcut: "n", action: { overlay.showMessage("Use Claude to create issues (coming soon)") }),
                        OverlayItem(label: "", dimmed: true),
                        OverlayItem(label: "[C] Reconfigure Tracker", shortcut: "c", action: {
                            showTrackerSetup(overlay: overlay, config: config, onConfigured: { onTrackerRefresh(); onUpdate() }, onUpdate: onUpdate)
                        }),
                    ])
                    return
                }

                var items: [OverlayItem] = []

                if let current = statusBar.currentIssue {
                    items.append(OverlayItem(label: "Current: \(current.key) \u{2014} \(current.title)", value: current.status, dimmed: true))
                    items.append(OverlayItem(label: "", dimmed: true))
                }

                for issue in issues {
                    let isCurrent = statusBar.currentIssue?.key == issue.key
                    let icon = priorityToIcon(issue.priority)
                    let label = "\(isCurrent ? "\u{25c6}" : " ") \(issue.key)  \(icon) \(truncate(issue.title, max: 40))"
                    items.append(OverlayItem(label: label, value: issue.status, action: {
                        selectIssue(overlay: overlay, tracker: tracker, session: session, statusBar: statusBar, issue: issue, config: config, onTrackerRefresh: onTrackerRefresh, onUpdate: onUpdate)
                    }))
                }

                items.append(OverlayItem(label: "", dimmed: true))
                items.append(OverlayItem(label: "[N] Create New Issue", shortcut: "n", action: { overlay.showMessage("Use Claude to create issues (coming soon)") }))
                items.append(OverlayItem(label: "[C] Reconfigure Tracker", shortcut: "c", action: {
                    showTrackerSetup(overlay: overlay, config: config, onConfigured: { onTrackerRefresh(); onUpdate() }, onUpdate: onUpdate)
                }))

                overlay.updateItems(items)
            }
        } catch {
            await MainActor.run {
                overlay.setLoading(false)
                overlay.updateItems([
                    OverlayItem(label: "Error: \(error.localizedDescription)", dimmed: true),
                    OverlayItem(label: "", dimmed: true),
                    OverlayItem(label: "[R] Retry", shortcut: "r", action: {
                        loadIssues(overlay: overlay, tracker: tracker, session: session, statusBar: statusBar, config: config, onTrackerRefresh: onTrackerRefresh, onUpdate: onUpdate)
                    }),
                    OverlayItem(label: "[C] Reconfigure Tracker", shortcut: "c", action: {
                        showTrackerSetup(overlay: overlay, config: config, onConfigured: { onTrackerRefresh(); onUpdate() }, onUpdate: onUpdate)
                    }),
                ])
            }
        }
    }
}

private func selectIssue(
    overlay: OverlayManager,
    tracker: TrackerService,
    session: ClaudeSession,
    statusBar: StatusBarState,
    issue: Issue,
    config: Config,
    onTrackerRefresh: @escaping () -> Void,
    onUpdate: @escaping () -> Void
) {
    overlay.updateItems([
        OverlayItem(label: "\(issue.key): \(issue.title)", dimmed: true),
        OverlayItem(label: "Status: \(issue.status)  Priority: \(issue.priority)", dimmed: true),
        OverlayItem(label: "", dimmed: true),
        OverlayItem(label: "[C] Set as Current Issue", shortcut: "c", action: {
            statusBar.currentIssue = issue
            onUpdate()
            overlay.showMessage("Set current: \(issue.key)")
        }),
        OverlayItem(label: "[A] Analyze Issue", shortcut: "a", action: {
            statusBar.currentIssue = issue
            session.write("Analyze this issue and tell me what you think the problem is and what files are involved:\n\(issue.key): \(issue.title)\n")
            overlay.close()
        }),
        OverlayItem(label: "[P] Plan Implementation", shortcut: "p", action: {
            statusBar.currentIssue = issue
            session.write("Create an implementation plan for this issue. List the files to change and the approach:\n\(issue.key): \(issue.title)\n")
            overlay.close()
        }),
        OverlayItem(label: "[X] Fix Issue", shortcut: "x", action: {
            statusBar.currentIssue = issue
            session.write("Fix this issue. Analyze it, plan the fix, implement it, and test it:\n\(issue.key): \(issue.title)\n")
            overlay.close()
        }),
        OverlayItem(label: "[S] Update Status", shortcut: "s", action: {
            showStatusPicker(overlay: overlay, tracker: tracker, issue: issue, onUpdate: onUpdate)
        }),
        OverlayItem(label: "[O] Open in Browser", shortcut: "o", action: {
            NSWorkspace.shared.open(URL(string: issue.url)!)
            overlay.showMessage("Opening in browser...")
        }),
        OverlayItem(label: "", dimmed: true),
        OverlayItem(label: "[B] \u{2190} Back to List", shortcut: "b", action: {
            loadIssues(overlay: overlay, tracker: tracker, session: session, statusBar: statusBar, config: config, onTrackerRefresh: onTrackerRefresh, onUpdate: onUpdate)
        }),
    ])
}

private func showStatusPicker(
    overlay: OverlayManager,
    tracker: TrackerService,
    issue: Issue,
    onUpdate: @escaping () -> Void
) {
    let statuses = tracker.type == "linear"
        ? ["Backlog", "Todo", "In Progress", "In Review", "Done", "Canceled"]
        : ["To Do", "In Progress", "In Review", "Done"]

    overlay.updateItems(statuses.map { s in
        OverlayItem(label: "\(s == issue.status ? "\u{25c6} " : "  ")\(s)", action: {
            overlay.setLoading(true)
            Task {
                do {
                    try await tracker.updateStatus(key: issue.key, status: s)
                    await MainActor.run {
                        overlay.showMessage("Updated to \(s)")
                        onUpdate()
                        overlay.setLoading(false)
                    }
                } catch {
                    await MainActor.run {
                        overlay.showMessage("Error: \(error.localizedDescription)")
                        overlay.setLoading(false)
                    }
                }
            }
        })
    })
}

private func priorityToIcon(_ priority: String) -> String {
    let p = priority.lowercased()
    if p.contains("urgent") || p.contains("critical") { return "\u{25a0}\u{25a0}\u{25a0}" }
    if p.contains("high") { return "\u{25a0}\u{25a0}\u{25a1}" }
    if p.contains("medium") || p.contains("med") { return "\u{25a0}\u{25a1}\u{25a1}" }
    if p.contains("low") { return "\u{25a1}\u{25a1}\u{25a1}" }
    return "\u{b7}\u{b7}\u{b7}"
}

private func truncate(_ str: String, max: Int) -> String {
    str.count > max ? String(str.prefix(max - 1)) + "\u{2026}" : str
}

import AppKit
